### ISSUE TITLE
File upload randomly crashes

### ISSUE THREAD

**Issue by ErlendFax** [googleapis/nodejs-storage#1623]
The bug report won't be as thorough as I hoped and since this happens so rarely and without any notice we haven't been able to reproduce this locally - sorry about that. 

We have a service that runs a job every hour. It reads a DB and creates CSV files, and uploads them to `Google Cloud Storage`. Each job saves a few thousand files at about 150 KB each. A few times a month or so it _randomly crashes_ and in our logs we find two different errors. These does not occur at the same time so might not be related, but both of them happens "randomly" and both errors include an unhandled error event in `Duplexify`.

Missing a file and retrying is fine, but the crash takes down our pod and makes us loose a lot of files. 

The unhandled error starts in `file.js` in `createWriteStream()` when calling `fileWriteStream.destroy(error)`.

I might be wrong there, but both of these would normally fail gracefully, but the unhandled error in `Duplexify` makes them crash? Would fixing this be a solution, or should one look further into the reason for destroying at the first place? Has anyone experienced similar behaviour? 

### Errors

```
events.js:292
 throw er; // Unhandled 'error' event
 ^

RequestError: The uploaded data did not match the data from the server. As a precaution, the file has been deleted. To be sure the content is the same, you should try uploading the file again. at ... /node_modules/@google-cloud/storage/build/src/file.js:1336:35 at ... /node_modules/@google-cloud/common/build/src/util.js:367:25 at Util.handleResp (... /node_modules/@google-cloud/common/build/src/util.js:144:9) at ... /node_modules/@google-cloud/common/build/src/util.js:432:22 at onResponse (... /node_modules/retry-request/index.js:214:7) at ... /node_modules/teeny-request/build/src/index.js:219:17 at runMicrotasks (<anonymous>) at processTicksAndRejections (internal/process/task_queues.js:97:5)
Emitted 'error' event on Pumpify instance at:
 at errorOrDestroy (internal/streams/destroy.js:108:12)
 at Pumpify.onerror (_stream_readable.js:753:7)
 at Pumpify.emit (events.js:315:20)
 at Pumpify.Duplexify._destroy (... /node_modules/pumpify/node_modules/duplexify/index.js:195:15)
 at ... /node_modules/pumpify/node_modules/duplexify/index.js:185:10
 at processTicksAndRejections (internal/process/task_queues.js:79:11) {
 code: 'FILE_NO_UPLOAD',
 errors: [ null ]
}
```

```
events.js:292
 throw er; // Unhandled 'error' event
 ^

Error: Upload failed at Upload.<anonymous> (... /node_modules/gcs-resumable-upload/build/src/index.js:188:30) at Upload.emit (events.js:327:22) at Upload.onResponse (... /node_modules/gcs-resumable-upload/build/src/index.js:383:14) at Upload.makeRequestStream (... /node_modules/gcs-resumable-upload/build/src/index.js:332:14) at runMicrotasks (<anonymous>) at processTicksAndRejections (internal/process/task_queues.js:97:5) at async Upload.startUploading (... /node_modules/gcs-resumable-upload/build/src/index.js:216:13)
Emitted 'error' event on Pumpify instance at:
 at errorOrDestroy (internal/streams/destroy.js:108:12)
 at Pumpify.onerror (_stream_readable.js:753:7)
 at Pumpify.emit (events.js:315:20)
 at Pumpify.Duplexify._destroy (... /node_modules/pumpify/node_modules/duplexify/index.js:195:15)
 at ... /node_modules/pumpify/node_modules/duplexify/index.js:185:10
 at processTicksAndRejections (internal/process/task_queues.js:79:11)

```

### Code:
```ts
  const filePath = "TEST/device/system/2021/01/01/09.csv"

  storeNewFile = async (
    file: File,
    data: Data[]
  ): Promise<void> => {
    const storage = new Storage( xxx );
    const bucket = storage.bucket("myBucket");
    const file = bucket.file(filePath);
    
    const csv : string = await convertJsonToCsv(JSON.stringify(data), true);

    return new Promise<void>((resolve, reject) => {
      const input = new Readable();
      input.pipe(file.createWriteStream());

      input.on("end", resolve);
      input.on("error", reject);

      input.push(csv);
      input.push(null);
    });
  };
```

#### Environment details

  - OS: node:12-alpine docker image (our prod environment)
  - Node.js version: 12.22.6
  - npm version: 6.14.15
  - `@google-cloud/storage` version: 4.7.0
  - Typescript: 3.7.3

EDIT: I posted wrong code snippet. We do not append to any existing files, but create a new ones.

**Comment by ddelgrosso1** [comment]
Hi @ErlendFax thanks for opening this issue. A few questions for you, is there any kind of correlation between the crash and the data in the files? Is it at all possible that these files are some how getting corrupted? When you say this only happens a few times a month what would you say is typical as far as the number of times?

**Comment by ErlendFax** [comment]
Hi @ddelgrosso1, and thank you for your reply! Not that we can find. The file is just a text string in our ts program before storing it with an csv extension in Storage (check updated code snippet). This usually happens about 5-10 times a month.

Sorry for not being able to reproduce this, I'll try to run it locally for a while to see if it crashes here. Was hoping someone had an idea. 

Also, for us at least, I believe the crash is the biggest problem. We are able to reproduce the crash in the `Duplexify` package (see issue mentioned above), if it's of any interest.

**Comment by ddelgrosso1** [comment]
Thanks for the additional information @ErlendFax. I will do some investigation / experimenting on our end. Have you seen this behavior with any other versions of the storage library or have you consistently been on `4.7.0`?

**Comment by ErlendFax** [comment]
Thank you! We will keep looking ourself of course. `4.7.0` only, for 1.5 year.

**Comment by surjikal** [comment]
@erlendfax Hopefully this is not terrible advice, but for production on v4, we've had luck dealing with this sort of issue by:

1. Using the `{resumable: false}` option on the writestream. This should work well for you since you're pushing data in one big chunk, and you're dealing with small-ish data (e.g. small enough to not blow up on `JSON.stringify`).

2. I've found it necessary to implement my own retry logic on top of the library.

I retry, with exponential backoff, whenever I get `error.code` =
- ETIMEDOUT
- ECONNRESET
- greater than or equal to 500
- 408
- 429
- basically any other error that's not between 400 and 405

**Comment by ErlendFax** [comment]
Thank you @surjikal, will take a look.

You say v4, could updating to v5 of nodejs storage help? I couldn't find any issues fixing this in newer versions so didn't think updating would help.

**Comment by ddelgrosso1** [comment]
@ErlendFax if the issue is coming from `Duplexify` v5 of the storage library is still using the same versioning. However, there have been many improvements in v5 with regards to retrying failed operations that would alleviate a lot of having to handle it yourself.

**Comment by ddelgrosso1** [comment]
Hi @ErlendFax I just wanted to check back to see if trying v5 of the library was an option for you?

**Comment by ErlendFax** [comment]
Thank you for checking up on us @ddelgrosso1! Much appreciated.

I've added more logging, `resumable: false`, and upgraded to latest version. Its been running a few days in prod and no issues so far. However, we had one incident in our dev environment an hour after initial deploy. I'll post the error log below. Since this has only happened once in dev I don't really wan't to draw any conclusions yet. 

I'll let the prod service run for a while and come back to you with a status. 

```
events.js:291
 throw er; // Unhandled 'error' event
 ^
Api: Service Unavailable at new Api (/app/packages/ourService/node_modules/@google-cloud/common/build/src/util.js:73:15) at Util.parseHttpRespMessage (/app/packages/ourService/node_modules/@google-cloud/common/build/src/util.js:175:41) at Util.handleResp (/app/packages/ourService/node_modules/@google-cloud/common/build/src/util.js:149:76) at /app/packages/ourService/node_modules/@google-cloud/common/build/src/util.js:477:22 at onResponse (/app/packages/ourService/node_modules/retry-request/index.js:228:7) at /app/packages/ourService/node_modules/teeny-request/build/src/index.js:157:17 at processTicksAndRejections (internal/process/task_queues.js:97:5)
Emitted 'error' event on Pumpify instance at:
 at errorOrDestroy (internal/streams/destroy.js:108:12)
 at Pumpify.onerror (_stream_readable.js:754:7)
 at Pumpify.emit (events.js:314:20)
 at Pumpify.Duplexify._destroy (/app/packages/ourService/node_modules/pumpify/node_modules/duplexify/index.js:195:15)
 at /app/packages/ourService/node_modules/pumpify/node_modules/duplexify/index.js:185:10
 at processTicksAndRejections (internal/process/task_queues.js:79:11) {
 code: 503,
 errors: [],
 response: PassThrough {
 _readableState: ReadableState {
 objectMode: false,
 highWaterMark: 16384,
 buffer: BufferList { head: null, tail: null, length: 0 },
 length: 0,
 pipes: null,
 pipesCount: 0,
 flowing: true,
 ended: true,
 endEmitted: true,
 reading: false,
 sync: false,
 needReadable: false,
 emittedReadable: false,
 readableListening: false,
 resumeScheduled: false,
 emitClose: true,
 autoDestroy: false,
 bufferedRequestCount: 0,
 corkedRequestsFree: {
 next: null,
 entry: null,
 finish: [Function: bound onCorkedFinish]
 }
 },
 writable: false,
 allowHalfOpen: true,
 _transformState: {
 afterTransform: [Function: bound afterTransform],
 needTransform: false,
 transforming: false,
 writecb: null,
 writechunk: null,
 writeencoding: 'buffer'
 },
 statusCode: 503,
 statusMessage: 'Service Unavailable',
 request: {
 agent: Agent {
 _events: [Object: null prototype] {
 free: [Function],
 newListener: [Function: maybeEnableKeylog]
 },
 _eventsCount: 2,
 _maxListeners: undefined,
 defaultPort: 443,
 protocol: 'https:',
 options: { maxSockets: Infinity, keepAlive: true, path: null },
 requests: {},
 sockets: {},
 freeSockets: {
 'storage.googleapis.com:443::::::::::::::::::': [ [TLSSocket], [TLSSocket] ]
 },
 keepAliveMsecs: 1000,
 keepAlive: true,
 maxSockets: Infinity,
 maxFreeSockets: 256,
 maxTotalSockets: Infinity,
 totalSocketCount: -19,
 scheduling: 'fifo',
 maxCachedSessions: 100,
 _sessionCache: { map: {}, list: [] },
 [Symbol(kCapture)]: false
 },
 headers: {
 'User-Agent': 'gcloud-node-storage/5.14.4',
 'x-goog-api-client': 'gl-node/12.22.6 gccl/5.14.4',
 Authorization: 'Bearer xxxx',
 'Content-Type': 'multipart/related; boundary=3b164dc2-4444-4444-99f5-9b43d04de389'
 },
 href: 'https://storage.googleapis.com/upload/storage/v1/b/ourBucket/o?uploadType=multipart&name=pikkM8SV-wM%2FOyBZdwI0u.csv'
 },
 body: 'Service Unavailable',
 headers: {
 'content-length': '19',
 'content-type': 'text/plain; charset=utf-8',
 date: 'Sat, 02 Oct 2021 02:08:43 GMT',
 server: 'UploadServer',
 'x-guploader-uploadid': 'xxx'
 },
 toJSON: [Function: toJSON],
 [Symbol(kCapture)]: false
 }
}
```

**Comment by surjikal** [comment]
@ErlendFax You need to implement the retry handling on your side for `resumable: false` for v5, at least as of right now. If I were you I'd try `file.save` since your sample code is very similar to how it's implement. I'm just not sure if it retries after hitting your original error though.

**Comment by ErlendFax** [comment]
Thanks, rewrote it to use `file.save` now. 
```ts
storeNewFile = async (
    file: File,
    data: Data[]
  ): Promise<void> => {
    const csv = await convertJsonToCsv(JSON.stringify(data), true);
    this.logger.info(
      `Storing new file ${file.name} of string length: ${csv.length}`
    );
    await file.save(csv);
```


> I'm just not sure if it retries after hitting your original error though.

Yeah we have I feeling it might not, but lets see what happens now after upgrading to v5

**Comment by ddelgrosso1** [comment]
I was mistaken, `createWriteStream` will not retry. If retries are needed please use the `save` function which ultimately will call `createWriteStream`.

**Comment by ddelgrosso1** [comment]
Hi @ErlendFax, I wanted to check in again and see how the stability has been since upgrading to v5 / using `save`? Have you been able to further reproduce the problem you were seeing previously?

**Comment by ErlendFax** [comment]
@ddelgrosso1 Hi! As mentioned in my first post, these errors usually happened a few times a months and sometimes with a few weeks between errors. I think we should let it run a few weeks before we draw any conclusions, but no errors so far. Looks good for now. 

Is that OK for you?

**Comment by shaffeeullah** [comment]
@ErlendFax , 

Resumable upload URI creation retries are now implemented (as of `5.15.6`). We suspect this was your issue. Does upgrading to this version resolve the crashing?

**Comment by ErlendFax** [comment]
Thanks, updated package now. I'll let you know if it crashes again within the coming week.

**Comment by shaffeeullah** [comment]
@ErlendFax any update?

**Comment by ErlendFax** [comment]
No crashes so far! Big thanks to all of you 💯 

However, it's kinda slow, about three times as slow as before all of this. Any suggestions?

**Comment by shaffeeullah** [comment]
Woo! 🎉  That's amazing to hear. It is taking longer because instead of crashing on a failed request, it is retrying the request using exponential backoff. Retry behavior and more information on exponential backoff are documented on [this page](https://cloud.google.com/storage/docs/retry-strategy#exponential-backoff).

**Comment by ErlendFax** [comment]
Thanks! This is kinda drifting off its topic, but I gotta ask. When looking at the logs it seems like _all_ of the uploads now use _at least_ three seconds to upload a few kb. Anything we can do to speed it up a bit?

**Comment by ddelgrosso1** [comment]
@ErlendFax would you happen to know what sort of times you were seeing previously? Do you know what percentage of the uploads are being retried currently?

**Comment by ErlendFax** [comment]
Yeah, streams are now considerably slower. 

Before I upgraded an upload would take 300-500ms, now it's usually 3-4 secs. I have no logs of the actual retries (is it possible?). 

From what I can see from the docs the initial wait time is 1 sec, those are not possible for me to spot right now, they don't differ much from regular upload times. The ones with longer uploading times are easier to spot, there are less than 10 uploads taking more than 5 sec of about 1000 files - not possible to say if these are actual retries or just slow internet speeds in our gcp k8s cluster. The longer waits, 10+ sec, are rare. Once every 1000-2000 files or so. 

Let me know if you want me to test anything.

**Comment by ddelgrosso1** [comment]
Thank you for the information @ErlendFax. I will run some experiments on our side and take a look at upload times pre/post changes. I may ultimately close out this issue and open another as it appears the original problem has been resolved.

**Comment by ErlendFax** [comment]
I see now that the slow upload speeds are caused by a memory leak in our cluster. The upload speed on a fresh cluster is good (<1sec). We use `node-cron` to run an hourly job and when _not_ adding `resumable: false` it slowly eats up the memory - when setting it to `false` we have no leak. 

I read somewhere that there is a file used internally for resumable uploads, it stores upload info in case of connectivity issues, maybe it grows really big and never gets deleted? `Resumable: false` will probably disable retries? Idk ... 

Anyways, feel free to mark it as resolved. Thank you for your help!

**Comment by ddelgrosso1** [comment]
Thank you for the followup. You are correct, under the hood the `gcs-resumable-upload` library does use `configstore` which utilizes the local file system. There is an issue [here](https://github.com/googleapis/gcs-resumable-upload/issues/347) to potentially modify that behavior. This would be the first I have heard about a possible leak so if you want to chime in on that issue with any information that would be great. In the meantime, going to close this out.

**Comment by nguyenphuocnhatthanh** [comment]
I have same error right now (Error 408 (Request Timeout)).
`return new Promise<void>((resolve, reject) => {
      const input = new Readable();
      const file = bucket.file(path);
      input.pipe(file.createWriteStream({     resumable: false }));
      input.on("end", resolve);
      input.on("error", function () {
        console.log("Oh no!");  // <------
        reject();
      });
      input.push(csv);
      input.push(null);
    });`
    My code is similar above. The error sometimes happend in google cloud function but it always work well in my local. I dont know the reason.
    My version package is 5.16.0. 
    Could you please help me check this one? @ErlendFax  @ddelgrosso1

**Comment by ddelgrosso1** [comment]
Hi @nguyenphuocnhatthanh I just want to confirm, are you seeing a crash happening or are you getting an HTTP 408 error? If you would like, it is probably best to open a new issue for the problem you are encountering unless you feel it is somehow related to this issue? Thanks!

**Comment by nguyenphuocnhatthanh** [comment]
I find out my issue that is relate my sequence code in cloud function. I changed flow response and it works well. Thank your reply @ddelgrosso1 .
